consensus/core/src/dag_state.rs

```diff
@@ -233,6 +234,40 @@ impl DagState {
             );
         }
 
+        if state.gc_enabled() {
+            if let Some(last_commit) = last_commit {
+                let mut index = last_commit.index();
+                let gc_round = state.gc_round();
+                info!("Recovering block commit statuses from commit index {} and backwards until leader of round <= gc_round {:?}", index, gc_round);
+
+                loop {
+                    let commits = store
+                        .scan_commits((index..=index).into())
+                        .unwrap_or_else(|e| panic!("Failed to read from storage: {:?}", e));
+                    let Some(commit) = commits.first() else {
+                        break;
+                    };
+
+                    // Check the commit leader round to see if it is within the gc_round. If it is not then we can stop the recovery process.
+                    if gc_round > 0 && commit.leader().round <= gc_round {
+                        break;
+                    }
+
+                    commit.blocks().iter().filter(|b| b.round > gc_round).for_each(|block_ref|{
+                        debug!(
+                            "Setting block {:?} as committed based on commit {:?}",
+                            block_ref,
+                            commit.index()
+                        );
+                        assert!(state.set_committed(block_ref), "Attempted to set again a block {:?} as committed when recovering commit {:?}", block_ref, commit);
+                    });
+
+                    // When reaching zero will exist anyways as all commits are indexed starting from 1.
+                    index = index.saturating_sub(1);
+                }
+            }
```

Comment:
Alternatively we could always take all the committed sub dag with index range `(last_commit_index - gc_depth...=last_commit_index)` to do the restore of commit status for the blocks above the gc_round. This would work quite well without having to do any iterations , although the reason why this is correct and we have strong guarantees that we restore enough data is less obvious and maybe for error prone in the future on refactorings. Instead I chose the above approach which is more straightforward to reasonate about. Performance gains most probably won't be noticeable.

Reply:
I think you should just read from store in range and then do the same processing in memory.